game-core/src/main/java/games/strategy/engine/data/RouteFinder.java

```diff
@@ -66,19 +65,65 @@
     return Optional.empty();
   }
 
+  Optional<Route> findRouteByCost(final Territory start, final Territory end) {
```

Comment:
This is the core and most important method of the PR. I'd appreciate focused review here and anything that can help with performance.

Reply:
As a first observation, there could be more abstraction or decomposition. Just noting we've maps and number types and not too much that is higher order.

For example, I suspect `final Map<Territory, BigDecimal> routeCosts = new HashMap<>();` could be replaced with a `CostCalculator` and then hide a number of details and be more easily tested on its own.

You could have methods like "addCost", or "compareTerritory", or "isBetterRoute", seems like a number of the logic conditionals in this method could be folded into it.

re: performance, @ron-murhammer Do you know the method call frequency or current runtime?

Reply:
This is part of route finding so every time you highlight units and move the cursor around. @RoiEXLab did some performance testing when he rewrote the route finding algorithm. I haven't run any numbers yet just manual testing on maps so far.

Reply:
Hmm the code doesn't really look like a common Dijkstra algorithm, which may indicate the performance isn't optimal.
Might be worth going for A* even if the performance is bad

Reply:
@RoiEXLab Yeah, its more of an adaption to the existing route finding. My guess is its not optimal but performance seems ok. I figure we can also always come back and refactor it since its very contained. Open to help here :)

Reply:
Agree. I was waiting to get rid of the old implementation until after I was finished and tested performance. Performance comparison was pretty negligible for route finding so I've combined them now.